### PR TITLE
Fix non-deterministic ordering of quick actions

### PR DESCRIPTION
**Pull request by mavasani**
**Note**: Recommended to review ignoring whitespace changes

https://github.com/dotnet/roslyn/commit/415d5101e8d524354bad5915d2850044ea6ec196 expanded tracking pending action sets to all priority buckets. However, it switched from using an ArrayBuilder to a MultiDictionary whose values are sets. This means the original order of pending actions for a specific bucket is lost.

This PR switches from a MultiDictionary to a regular Dictionary with ArrayBuilder values. This fixes the non-determinism in the ordering.

**Testing**: I have also added an integration test to guard against this regression. Verified that this integration test fails prior to the product fix.

### CODE REVIEW

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 1]
```diff
@@ -1458,5 +1458,61 @@ public async Task TestNonSourceDocumentRefactoring()
...
+                "Wrap every parameter",
+                "Align wrapped parameters",
+                "Indent all parameters",
+                "Indent wrapped parameters",
+                "Unwrap and indent all parameters",
+            };
+
+            await TestServices.EditorVerifier.CodeActionsAsync(expectedItems, ensureExpectedItemsAreOrdered: true, cancellationToken: HangMitigatingCancellationToken);
+
+            await SetUpEditorAsync(@"
```
Comment (mavasani) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:1492]: Added 2 invocations as it seems the first one might succeed often, but for 9 out of 10 times I saw either first or second one has unexpected order of the items.

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs [hunk 2]
```diff
@@ -1458,5 +1458,61 @@ public async Task TestNonSourceDocumentRefactoring()
...
+#pragma warning disable IDE0060 // Remove unused parameter
+class C
+{ 
+    public C(int x1, int $$x2, int x3)
+    {
+    }
+}", HangMitigatingCancellationToken);
+
+            await TestServices.Editor.InvokeCodeActionListAsync(HangMitigatingCancellationToken);
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpCodeActions.cs:1474]: Good question- I saw the test running did perform two lightbulb invocations at the locations I expected. Maybe this call is not needed after all. I’ll investigate and cleanup in a follow up PR tomorrow!

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource_Async.cs [hunk 3]
```diff
@@ -102,79 +102,96 @@ private partial class SuggestedActionsSource : IAsyncSuggestedActionsSource
...
+
+                                    // if the actual priority class is lower than the one we're currently in, then hold onto
+                                    // this set for later, and place it in that priority group once we get there.
+                                    if (actualSetPriority < priority)
+                                    {
+                                        if (!pendingActionSets.TryGetValue(actualSetPriority, out var builder))
+                                        {
+                                            builder = ArrayBuilder<SuggestedActionSet>.GetInstance();
+                                            pendingActionSets.Add(actualSetPriority, builder);
+                                        }
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource_Async.cs:148]: 💡 The `GetOrAdd` helper makes this easier

```suggestion
                                        var builder = pendingActionSets.GetOrAdd(actualSetPriority, _ => ArrayBuilder<SuggestedActionSet>.GetInstance());
```

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource_Async.cs [hunk 4]
```diff
@@ -102,79 +102,96 @@ private partial class SuggestedActionsSource : IAsyncSuggestedActionsSource
...
-
-                        // Ensure we always complete the collector even if we didn't add any items to it.
-                        // This ensures that we unblock the UI from displaying all the results for that
-                        // priority class.
-                        collector.Complete();
-                        completedCollectors.Add(collector);
+                    }
+                    finally
+                    {
+                        foreach (var builder in pendingActionSets.Values)
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource_Async.cs:193]: 💡 The enumerator for the dictionary is more efficient than going through `Values`.

```suggestion
                        foreach (var (_, builder) in pendingActionSets)
```

**Review by sharwell — APPROVED** [review]

**Pull request merged**